x/staking/keeper/msg_server.go

```diff
@@ -0,0 +1,257 @@
+package keeper
+
+import (
+	"bytes"
+	"context"
+	"math/big"
+	"strconv"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	hmCommon "github.com/maticnetwork/heimdall/common"
+	"github.com/maticnetwork/heimdall/helper"
+	hmTypes "github.com/maticnetwork/heimdall/types"
+	"github.com/maticnetwork/heimdall/x/staking/types"
+)
+
+type msgServer struct {
+	Keeper
+	contractCaller helper.IContractCaller
+}
+
+// NewMsgServerImpl returns an implementation of the bank MsgServer interface
+// for the provided Keeper.
+func NewMsgServerImpl(keeper Keeper, contractCaller helper.IContractCaller) types.MsgServer {
+	return &msgServer{Keeper: keeper, contractCaller: contractCaller}
+}
+
+var _ types.MsgServer = msgServer{}
+
+func (k msgServer) ValidatorJoin(goCtx context.Context, msg *types.MsgValidatorJoin) (*types.MsgValidatorJoinResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	k.Logger(ctx).Debug("✅ Validating validator join msg",
+		"validatorId", msg.ID,
+		"activationEpoch", msg.ActivationEpoch,
+		"amount", msg.Amount,
+		"SignerPubkey", msg.SignerPubKey.String(),
+		"txHash", msg.TxHash,
+		"logIndex", msg.LogIndex,
+		"blockNumber", msg.BlockNumber,
+	)
+
+	// Generate PubKey from Pubkey in message and signer
+	pubkey := msg.SignerPubKey
+	signer := pubkey.Address()
+
+	// Check if validator has been validator before
+	if _, ok := k.GetSignerFromValidatorID(ctx, msg.ID); ok {
+		k.Logger(ctx).Error("Validator has been validator before, cannot join with same ID", "validatorId", msg.ID)
+		return nil, hmCommon.ErrValidatorAlreadyJoined
+	}
+
+	// get validator by signer
+	checkVal, err := k.GetValidatorInfo(ctx, signer.Bytes())
```

Comment:
We have changed these methods in query.go to:
```
		GetValidatorInfoCmd(),
		GetCurrentValSetCmd(),

```
I think we can leave them as GetValidatorInfo() and GetCurrentValSet() so that we don't have to change it other places.